### PR TITLE
Initial support for strings

### PR DESCRIPTION
**Pull request by isubasinghe**
This PR adds support for string slices (&str) and strings (String), however due to not being able to attach specifications to &str yet, new types (the StrSlice and String structs) are used as a temporary solution. 

StrSlice currently supports 4 operations:
  * StrSlice.len()
  * StrSlice.get_char(int) -> u8
  * StrSlice.is_ascii() -> bool
  * StrSlice.substring(int, int) -> StrSlice

### CODE REVIEW

**Comment by isubasinghe** [comment]
Note intent to squash and merge

File: source/vir/src/context.rs [hunk 1]
```diff
@@ -70,6 +71,11 @@ pub struct Ctx {
     pub(crate) debug: bool,
     pub fun: Option<FunctionCtx>,
     pub global: GlobalCtx,
+    // In the very unlikely case where we get sha512 collisions
+    // we use this to panic rather than introduce unsoundness.
+    // Of course it can be argued that accounting for sha512 collisions
+    // is overkill, perhaps this should be revisited.
+    pub(crate) string_hashes: RefCell<HashMap<BigUint, Arc<String>>>,
```
Comment (isubasinghe) [inline source/vir/src/context.rs:78]: Is this unnecessary? a sha512 collision is extremely unlikely and even feels a bit silly to account for

File: source/vir/Cargo.toml [hunk 2]
```diff
@@ -10,6 +10,8 @@ edition = "2018"
 [dependencies]
 air = { path = "../air" }
 sise = "0.6.0"
+sha2 = "0.10.2"
+num-bigint = "0.4.3"
```
Comment (isubasinghe) [inline source/vir/Cargo.toml:14]: The hashing technique introduces two dependencies but the alternative of using a counter to assign an id to a string could be non deterministic. 

Perhaps the middle ground could be to use a 64 bit (or perhaps even 128 is fine) hash and then increment the integer until a collision is resolved.

File: source/vir/src/triggers_auto.rs [hunk 3]
```diff
@@ -335,6 +335,10 @@ fn gather_terms(ctxt: &mut Ctxt, ctx: &Ctx, exp: &Exp, depth: u64) -> (bool, Ter
                 )),
             )
         }
+        ExpX::Str(strop) => {
+            //TODO: visit this
+            todo!();
```
Comment (isubasinghe) [inline source/vir/src/triggers_auto.rs:340]: This doesn't seem to ever be triggered

Comment (isubasinghe) [reply source/vir/src/triggers_auto.rs:340]: I will hold off from opening a proper PR until this is resolved

File: source/rust_verify/tests/strings.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
+#![feature(rustc_private)]
```
Comment (utaal) [inline source/rust_verify/tests/strings.rs:1]: Please add tests for:
* inequality and equality of string literals, both as they appear inline and as part of spec functions or e.g. postconditions of functions, i.e.
```
fn hello_str<'a>() -> (res: StrSlice<'a>)
  ensures res.substring(0, 5) === new_strlit("Hello")
```
and a function that uses the result of `hello_str`.

Comment (utaal) [reply source/rust_verify/tests/strings.rs:1]: Also, please add tests with multiple StrSlices and literals in the same program.

File: source/vir/src/prelude.rs [hunk 5]
```diff
@@ -193,6 +213,28 @@ pub(crate) fn prelude_nodes(config: PreludeConfig) -> Vec<Node> {
             :skolemid skolem_prelude_box_unbox_sint
         )))
 
+        (axiom (forall ((x [Poly])) (!
```
Comment (utaal) [inline source/vir/src/prelude.rs:216]: Add a header comment "String literals"

**Review by utaal — CHANGES_REQUESTED** [review]

File: source/vir/Cargo.toml (hunk 2 above)
Comment (isubasinghe) [reply source/vir/Cargo.toml:14]: num-bigint is already used in the interpreter so this dependency is fine, I have decided to proceed with the approach as a result.

File: source/rust_verify/src/rust_to_vir_expr.rs [hunk 6]
```diff
@@ -936,6 +948,83 @@ fn fn_call_to_vir<'tcx>(
...
+                let arg0 = args.first().unwrap();
+                let arg0 = expr_to_vir(bctx, arg0, ExprModifier::REGULAR)
+                    .expect("internal compiler error");
+                let arg1 = &args[1];
+                let arg1 = expr_to_vir(bctx, arg1, ExprModifier::REGULAR)
+                    .expect("internal compiler error");
+                Ok(mk_expr(ExprX::Str(StrOp::GetChar { strslice: arg0, index: arg1 })))
+            }
+            _ => panic!(
+                "Expected a call for builtin::strslice_is_ascii with one argument but did not receive it"
```
Comment (utaal) [inline source/rust_verify/src/rust_to_vir_expr.rs:993]: `strslice_get_char`?

File: source/rust_verify/src/rust_to_vir_expr.rs [hunk 7]
```diff
@@ -936,6 +948,83 @@ fn fn_call_to_vir<'tcx>(
...
+        return match &expr.kind {
+            ExprKind::Call(_, args) => {
+                assert!(args.len() == 1);
+                let arg0 = args.first().unwrap();
+                let arg0 = expr_to_vir(bctx, arg0, ExprModifier::REGULAR)
+                    .expect("internal compiler error");
+                Ok(mk_expr(ExprX::Str(StrOp::Len(arg0))))
+            }
+            _ => panic!(
+                "Expected a call for builtin::strslice_is_ascii with one argument but did not receive it"
```
Comment (utaal) [inline source/rust_verify/src/rust_to_vir_expr.rs:1008]: `strslice_len`?

File: source/rust_verify/src/rust_to_vir_expr.rs [hunk 8]
```diff
@@ -936,6 +948,83 @@ fn fn_call_to_vir<'tcx>(
...
+        return Ok(mk_expr(ExprX::RevealString(Arc::new(s.to_string()))));
+    }
+
+    if is_strslice_get_char {
+        return match &expr.kind {
+            ExprKind::Call(_, args) => {
+                assert!(args.len() == 2);
+                let arg0 = args.first().unwrap();
+                let arg0 = expr_to_vir(bctx, arg0, ExprModifier::REGULAR)
+                    .expect("internal compiler error");
```
Comment (utaal) [inline source/rust_verify/src/rust_to_vir_expr.rs:986]: Please briefly describe what's unexpected here, instead. "invalid parameter for strslice_get_char", for example.

File: source/rust_verify/src/rust_to_vir_func.rs [hunk 9]
```diff
@@ -147,19 +218,23 @@ pub(crate) fn check_item_fn<'tcx>(
             );
         }
 
-        let typ = mid_ty_to_vir(ctxt.tcx, is_mut.unwrap_or(input), false);
+        let typ: Typ = mid_ty_to_vir(ctxt.tcx, is_mut.unwrap_or(input), false);
```
Comment (utaal) [inline source/rust_verify/src/rust_to_vir_func.rs:221]: Why this change?

File: source/vir/Cargo.toml (hunk 2 above)
Comment (utaal) [reply source/vir/Cargo.toml:14]: 👍

File: source/vir/src/prune.rs [hunk 10]
```diff
@@ -103,6 +103,14 @@ fn traverse_reachable(ctxt: &Ctxt, state: &mut State) {
     loop {
         let ft = |state: &mut State, t: &Typ| {
             match &**t {
+                TypX::StrSlice => {
```
Comment (utaal) [inline source/vir/src/prune.rs:106]: Please add a note that this is temporary, until we support adding specification for `std`.

**Review by utaal — CHANGES_REQUESTED** [review]

File: source/rust_verify/src/rust_to_vir_func.rs (hunk 9 above)
Comment (isubasinghe) [reply source/rust_verify/src/rust_to_vir_func.rs:221]: I think  had some trouble with my LSP and was trying to figure out what was causing {unknown} types to appear, this was an artifact of my LSP debugging that I forgot to remove

File: source/rust_verify/src/rust_to_vir_expr.rs (hunk 8 above)
Comment (isubasinghe) [reply source/rust_verify/src/rust_to_vir_expr.rs]: Done

File: source/rust_verify/src/rust_to_vir_adts.rs [hunk 11]
```diff
@@ -93,18 +94,27 @@ pub fn check_item_struct<'tcx>(
...
     let name = hack_get_def_name(ctxt.tcx, def_id);
     let path = def_id_to_vir_path(ctxt.tcx, def_id);
+
     let variant_name = Arc::new(name.clone());
     let (variant, one_field_private) = if vattrs.external_body {
         (ident_binder(&variant_name, &Arc::new(vec![])), false)
     } else {
         let field_defs = adt_def.all_fields();
-        check_variant_data(ctxt, module_path, &variant_name, variant_data, false, field_defs)
+        check_variant_data(ctxt, &module_path, &variant_name, variant_data, false, field_defs)
```
[the remark this answers is not in the document]
Comment (isubasinghe) [reply source/rust_verify/src/rust_to_vir_adts.rs]: not sure myself, reverted it

**Comment by utaal** [comment]
I'm going to approve this, but please wait for @Chris-Hawblitzel's review before merging.

**Review by utaal — APPROVED** [review]
`pervasive::String` is barebones, but we can expand it in a future PR.

**Comment by utaal** [comment]
Ah, one more thing: can you rename `.view()` to `.bytes()` (or similar), so we can keep `.view()` for when we introduce a character representation of the string?

**Review by Chris-Hawblitzel — APPROVED** [review]
I just have a couple of comments to address, then go ahead and merge.

**Pull request merged**